### PR TITLE
Improve boost

### PR DESCRIPTION
**Pull request by NickeZ**
Don't merge this yet. I've implemented most of the things I want to, but it needs a cleanup and some more fixes.

Works for me on windows with three kinds of boost installations (from sf.net binaries, compiled as "layout=versioned" and "layout=system") and on ubuntu.

I tested by setting BOOST_ROOT to my different installations. It also picks up the most common locations by itself, such as c:\program files\boost, c:\boost, c:\local\boost_*.

### CODE REVIEW

File: mesonbuild/compilers/c.py [hunk 1]
```diff
@@ -1027,4 +1027,28 @@ def get_instruction_set_args(self, instruction_set):
             return None
         return vs32_instruction_set_args.get(instruction_set, None)
 
+    def get_toolset_version(self):
+	# See boost/config/compiler/visualc.cpp for up to date mapping
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/c.py]: __[flake8]__
* [E101] indentation contains mixed spaces and tabs
* [W191] indentation contains tabs

File: mesonbuild/compilers/c.py [hunk 2]
```diff
@@ -1027,4 +1027,28 @@ def get_instruction_set_args(self, instruction_set):
...
+        version = int(''.join(self.version.split('.')[0:2]))
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/c.py]: __[flake8]__
* [E101] indentation contains mixed spaces and tabs

**Comment by NickeZ** [comment]
I believe this is ready for review, all the tests passes. This is my first time contributing to meson so please look extra carefully.

**Comment by nirbheek** [comment]
The CI failures seem relevant to this PR, could you look into that?

**Comment by NickeZ** [comment]
Yeah you are right. At first I didn't see how they could relate, but looking at the python backtrace it made sense.

**Comment by NickeZ** [comment]
I don't understand why it fails on mac os. Travis doesn't provide any docs either. Maybe someone with a mac can run the tests?

**Comment by nirbheek** [comment]
Maybe you can recursively list the contents of `/usr/local/include/boost*` and see whether the header actually exists or not.

**Comment by NickeZ** [comment]
Maybe you can do a code review and we can find out if I'm doing something inherently wrong..

**Comment by NickeZ** [comment]
The failing test on windows was actually `test cases/common\\111 has header symbol`, with the following traceback:

```
Traceback (most recent call last):
  File "C:\projects\meson\mesonbuild\mesonmain.py", line 345, in run
    app.generate()
  File "C:\projects\meson\mesonbuild\mesonmain.py", line 140, in generate
    self._generate(env)
  File "C:\projects\meson\mesonbuild\mesonmain.py", line 188, in _generate
    intr.run()
  File "C:\projects\meson\mesonbuild\interpreter.py", line 2779, in run
    super().run()
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 148, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 170, in evaluate_codeblock
    raise e
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 164, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 185, in evaluate_statement
    return self.evaluate_if(cur)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 236, in evaluate_if
    self.evaluate_codeblock(i.block)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 170, in evaluate_codeblock
    raise e
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 164, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 185, in evaluate_statement
    return self.evaluate_if(cur)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 239, in evaluate_if
    self.evaluate_codeblock(node.elseblock)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 170, in evaluate_codeblock
    raise e
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 164, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 175, in evaluate_statement
    return self.function_call(cur)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 383, in function_call
    (posargs, kwargs) = self.reduce_arguments(node.args)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 538, in reduce_arguments
    reduced_pos = [self.evaluate_statement(arg) for arg in args.arguments]
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 538, in <listcomp>
    reduced_pos = [self.evaluate_statement(arg) for arg in args.arguments]
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 199, in evaluate_statement
    return self.evaluate_notstatement(cur)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 224, in evaluate_notstatement
    v = self.evaluate_statement(cur.value)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 179, in evaluate_statement
    return self.method_call(cur)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 413, in method_call
    return obj.method_call(method_name, self.flatten(args), kwargs)
  File "C:\projects\meson\mesonbuild\interpreterbase.py", line 96, in method_call
    return self.methods[method_name](args, kwargs)
  File "C:\projects\meson\mesonbuild\interpreter.py", line 963, in has_header_symbol_method
    haz = self.compiler.has_header_symbol(hname, symbol, prefix, self.environment, extra_args, deps)
  File "C:\projects\meson\mesonbuild\compilers\cpp.py", line 55, in has_header_symbol
    if super().has_header_symbol(hname, symbol, prefix, env, extra_args, dependencies):
  File "C:\projects\meson\mesonbuild\compilers\c.py", line 260, in has_header_symbol
    return self.compiles(t.format(**fargs), env, extra_args, dependencies)
  File "C:\projects\meson\mesonbuild\compilers\c.py", line 304, in compiles
    args = self._get_compiler_check_args(env, extra_args, dependencies, mode)
  File "C:\projects\meson\mesonbuild\compilers\c.py", line 275, in _get_compiler_check_args
    args += d.get_compile_args()
  File "C:\projects\meson\mesonbuild\dependencies\misc.py", line 205, in get_compile_args
    if not include_dir.endswith(('/usr/include', '/usr/local/include')):
AttributeError: 'NoneType' object has no attribute 'endswith'
```

It seems like the mac os build fails on the same test. So if someone has a mac, could you test without having boost installed? or just set `BOOST_ROOT=/garbage`

**Comment by NickeZ** [comment]
Could it be that `/usr/local/include` isn't included automatically on mac os?

**Comment by NickeZ** [comment]
Maybe there is a boost without math installed by default on the travis images?

**Comment by jpakkane** [comment]
Please rebase so travis will run the tests against current master. Reinvoking the build seems to build it against the old one.

**Comment by jpakkane** [comment]
`/usr/local/include` does not exist on OSX by default even, so it is unlikely to be in the search path by default. AFAICR using `/usr/local` is not recommended by Apple.

**Comment by NickeZ** [comment]
Where does brew install headers then? All I can find on the internet is /usr/local

**Comment by NickeZ** [comment]
Now it passes all the tests, I did a rebase to organize the commits a little bit. What do you think?

File: mesonbuild/compilers/c.py [hunk 3]
```diff
@@ -1027,4 +1027,28 @@ def get_instruction_set_args(self, instruction_set):
             return None
         return vs32_instruction_set_args.get(instruction_set, None)
 
+    def get_toolset_version(self):
+        # See boost/config/compiler/visualc.cpp for up to date mapping
+        version = int(''.join(self.version.split('.')[0:2]))
```
Comment (nirbheek) [inline mesonbuild/compilers/c.py]: This should be wrapped in a `try:` statement with an error if the version is unparseable. Can boost versions ever contain non-numeral characters? If so, this should use a sensible fallback instead.

File: mesonbuild/compilers/compilers.py [hunk 4]
```diff
@@ -1082,6 +1118,9 @@ def get_link_whole_for(self, args):
     def get_instruction_set_args(self, instruction_set):
         return gnulike_instruction_set_args.get(instruction_set, None)
 
+    def get_default_include_dirs(self):
+        return gnulike_default_include_dirs(self.exelist, self.language)
```
Comment (nirbheek) [inline mesonbuild/compilers/compilers.py]: This should also be copied into `IntelCompiler` since it has the same syntax as gcc and clang. Does it also make sense to implement this for MSVC? That would involve just reading the `INCLUDE` env var, I expect.

File: mesonbuild/compilers/compilers.py [hunk 5]
```diff
@@ -669,6 +670,9 @@ def get_option_compile_args(self, options):
     def get_option_link_args(self, options):
         return []
 
+    def get_default_include_dirs(self):
+        return []
```
Comment (nirbheek) [inline mesonbuild/compilers/compilers.py]: This should probably go on `CCompiler`, same as `get_library_dirs()`.

File: mesonbuild/dependencies/misc.py [hunk 6]
```diff
@@ -30,77 +30,156 @@
...
+
         self.lib_modules = {}
-        self.lib_modules_mt = {}
         self.detect_version()
-        self.requested_modules = self.get_requested(kwargs)
-        module_str = ', '.join(self.requested_modules)
         if self.is_found:
-            self.detect_src_modules()
             self.detect_lib_modules()
+            mlog.debug('Boost library directory is', self.libdir)
```
Comment (nirbheek) [inline mesonbuild/dependencies/misc.py]: Perhaps do `mlog.bold(self.libdir)`.

File: test cases/frameworks/1 boost/meson.build [hunk 7]
```diff
@@ -9,21 +9,18 @@ add_project_arguments(['-DBOOST_LOG_DYN_LINK'],
 # within one project. The need to be independent of each other.
 # Use one without a library dependency and one with it.
 
-nolinkdep = dependency('boost', modules: 'utility')
```
Comment (nirbheek) [inline test cases/frameworks/1 boost/meson.build]: Why was this removed?

File: mesonbuild/compilers/c.py (hunk 3 above)
Comment (NickeZ) [reply mesonbuild/compilers/c.py]: This is actually the mapping between the Visual Studio version like (14.0 or 14.1) and the compiler define _MSC_VER like 1900 and 1910. So it isn't boost related except that this is "experimentally found" by users of Visual Studio. There is no official mapping from MS. So it should always be numeric. I will add a try: block.

Comment (nirbheek) [reply mesonbuild/compilers/c.py]: Ah, makes sense.

File: mesonbuild/compilers/compilers.py (hunk 5 above)
Comment (NickeZ) [reply mesonbuild/compilers/compilers.py]: Ok

File: mesonbuild/compilers/compilers.py (hunk 4 above)
Comment (NickeZ) [reply mesonbuild/compilers/compilers.py]: Ok

File: mesonbuild/dependencies/misc.py (hunk 6 above)
Comment (NickeZ) [reply mesonbuild/dependencies/misc.py]: Sure

File: test cases/frameworks/1 boost/meson.build (hunk 7 above)
Comment (NickeZ) [reply test cases/frameworks/1 boost/meson.build]: Semantically there is no difference to depend on simply boost or depend on a "headers only library" which the `utility` module is.

File: mesonbuild/compilers/c.py [hunk 8]
```diff
@@ -1027,4 +1030,34 @@ def get_instruction_set_args(self, instruction_set):
...
+        elif version < 1900:
+            return '12.0' # (Visual Studio 2013)
+        elif version < 1910:
+            return '14.0' # (Visual Studio 2015)
+        elif version < 1920:
+            return '14.1' # (Visual Studio 2017)
+        return None
+
+    def get_default_include_dirs(self):
+        if not 'INCLUDE' in os.envvars:
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/c.py]: __[flake8]__
* [E713] test for membership should be 'not in'

**Comment by NickeZ** [comment]
Any more issues? Should I squash the last two commits into the correct commits?

**Comment by nirbheek** [comment]
Yes, please squash them. This LGTM for merging after that. :)

**Comment by NickeZ** [comment]
rebased

File: test cases/frameworks/1 boost/meson.build (hunk 7 above)
Comment (nirbheek) [reply test cases/frameworks/1 boost/meson.build]: Having a dependency that only involves using headers and no libraries is a good test to have. It ensures that our code doesn't assume that a dependency has both libs and includes. Does this not work on some platforms or something else?

Comment (NickeZ) [reply test cases/frameworks/1 boost/meson.build]: After my changes to the boost dependency this test is equal to the last test `nomoddep = dependency('boost')`.

Previously you could depend on for example `utility`. But today it will fail because `utility` is not a valid "module" any more. And it is not a valid module any more because it doesn't change anything. You will automatically get access to all the "header-only" boost modules because you only need -I/path/to/boost/include/folder to your build commands.

Comment (nirbheek) [reply test cases/frameworks/1 boost/meson.build]: Ah, I see. So it worked on accident earlier?

Comment (NickeZ) [reply test cases/frameworks/1 boost/meson.build]: No it was on purpose, it actually checked if for example `boost/utility.h` existed, but I don't think that makes sense because you cannot split up the boost headers. You always have to install all of them because they have very complex dependencies on each other. And it ended up confusing users because sometimes the header file and the library file has the same name and then their dependency could be reported as fine, but didn't actually work, because the header file existed but not the library. It is very common to only install a subset of the libraries.

Comment (nirbheek) [reply test cases/frameworks/1 boost/meson.build]: I think the new behaviour makes sense, but I don't know if it is acceptable to break existing build files which might be looking for the `utility` module, and so on. Perhaps it's fine since it's a backwards-compatible change. Let's go with that for now and revisit this if people complain.

**Comment by nirbheek** [comment]
Let's get this in, it looks great to me overall.

**Comment by NickeZ** [comment]
We should probably put some comment about the breakage in some release notes? The fix is pretty simply, just remove any modules that aren't found.

Since it worked even if you didn't specify the module before, I'm guessing that some people never did it in the first place.

**Comment by nirbheek** [comment]
Could you create a new directory `docs/markdown/Release-notes-for-0.43.0/` and create a file `001-boost-breakage.md` and put a subsection in there about this? See `Release-notes-for-0.42.0.md` for the format of subsections in the release notes. All files in that directory will be aggregated into `Release-notes-for-0.43.0.md` for the release.

**Comment by nirbheek** [comment]
On second thoughts it makes more sense to call it "Cross-platform Boost Dependency Improvements" and include a note about the header-only-modules change.

File: tools/boost_names.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    'multi_array': 'Multi-Array',
+    'multi_index': 'Multi-Index',
+    'numeric': 'Numeric Conversion',
+    'ptr_container': 'Pointer Container',
+    'poly_collection': 'PolyCollection',
+    'qvm': 'QVM',
+    'throw_exception': 'ThrowException',
+    'tti': 'TTI',
+    'vmd': 'VMD',
+    }
```
Comment (ignatenkobrain) [inline tools/boost_names.py]: __[flake8]__
* [E123] closing bracket does not match indentation of opening bracket's line

File: tools/boost_names.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    Module('numeric', 'Odeint', []),
+    Module('utility', 'Operators', []),
+    Module('core', 'Ref', []),
+    Module('utility', 'Result Of', []),
+    Module('algorithm', 'String Algo', []),
+    Module('core', 'Swap', []),
+    Module('', 'Tribool', []),
+    Module('numeric', 'uBLAS', []),
+    Module('utility', 'Value Initialized', []),
+    ]
```
Comment (ignatenkobrain) [inline tools/boost_names.py]: __[flake8]__
* [E123] closing bracket does not match indentation of opening bracket's line

File: tools/boost_names.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        else:
+            modname = dirname.replace('_', ' ').title()
+        modules.append(Module(dirname, modname, []))
+        sorted_modules = sorted(modules, key=lambda module: module.name.lower())
+
+    for module in sorted_modules:
+        print("{}: {}".format(module.dirname, module.name))
+
+
+if __name__=='__main__':
```
Comment (ignatenkobrain) [inline tools/boost_names.py]: __[flake8]__
* [E225] missing whitespace around operator

**Comment by NickeZ** [comment]
Writing the release notes had me thinking that maybe my approach isn't the correct one. Thinking from the users perspective we should probably use the official Boost module names and simply do nothing if the user specifies a header-only library.

I added a tool that correctly produces the correct module names. With some exceptions it is quite easy to map it from the directory names. 

What do you think about using a tool like this to produce a python map? Then we can choose if we either want to accept the directory names or proper boost module names or both.

My goal with the script is also to figure out which libraries actually are produced for each module by looking into the build files. Then we can use this information to better support cases like the "test" module which uses multiple libraries. If I have this information I can also correctly say that a module is installed or not.

**Comment by NickeZ** [comment]
I found that the human readable names actually are defined in json files.. Will rewrite my tool

File: mesonbuild/dependencies/misc.py [hunk 12]
```diff
@@ -30,77 +30,181 @@
...
+        compiler = ''
+        if mesonlib.is_windows():
+            suffix = '.dll'
+            if not self.static:
+                prefix = ''
+
+        return '{}{}{}{}{}{}'.format(prefix, libname_stem, realtime, debug, compiler, suffix)
+
+
+    def detect_nix_roots(self):
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[flake8]__
* [E303] too many blank lines (2)

File: mesonbuild/dependencies/misc.py [hunk 13]
```diff
@@ -245,33 +389,14 @@ def get_link_args(self):
...
             if default_detect is not None:
-                if module == 'unit_testing_framework':
-                    emon_args = self.compiler.find_library('boost_test_exec_monitor')
-                else:
-                    emon_args = None
                 args += default_detect
-                if emon_args is not None:
-                    args += emon_args
-            elif module in self.lib_modules or module in self.lib_modules_mt:
+            elif lib in self.lib_modules:
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[flake8]__
* [F821] undefined name 'lib'

File: mesonbuild/dependencies/misc.py [hunk 14]
```diff
@@ -671,3 +796,159 @@ def get_methods(self):
             return [DependencyMethods.PKGCONFIG, DependencyMethods.CUPSCONFIG, DependencyMethods.EXTRAFRAMEWORK]
         else:
             return [DependencyMethods.PKGCONFIG, DependencyMethods.CUPSCONFIG]
+
+
+# Generated with boost_names.py
+# (directory name, human readable name, library names)
+BOOST_MODULES = [('accumulators', 'Accumulators', []),
+ ('algorithm', 'Algorithm', []),
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[flake8]__
* [E128] continuation line under-indented for visual indent

File: tools/boost_names.py [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+
+                # Get metadata for module
+                jsonfile = os.path.join(root, f)
+                with open(jsonfile) as jsonfh:
+                    boost_modules = json.loads(jsonfh.read())
+                    if(isinstance(boost_modules, dict)):
+                        boost_modules = [boost_modules]
+                    for boost_module in boost_modules:
+                        modules.append(Module(boost_module['key'], boost_module['name'], libs))
+                        #print(boost_module['name'])
```
Comment (ignatenkobrain) [inline tools/boost_names.py]: __[flake8]__
* [E265] block comment should start with '# '

**Comment by nirbheek** [comment]
I like the 'ignore headers-only dependencies' approach too, and I like your idea of having a tool which we can re-run from time to time to validate/extend the list. 👍

**Comment by jpakkane** [comment]
This seems to have conflicts. Please rebase so the CI can get a fresh run.

**Comment by NickeZ** [comment]
This is "work in progress" since I decided to change approach.

**Comment by NickeZ** [comment]
Alright it seems like boost is a tricky dependency. I changed my opinion back to the breaking behaviour.

We have to leave it up to the build file writer to actually decide exactly which library to link against. For example the [test](http://www.boost.org/doc/libs/1_65_1/libs/test/doc/html/boost_test/usage_variants.html) module provides three libraries. But you can choose to not link at all. link statically or link shared to a single one of them or to any combination. 

Another example is the [stacktrace](http://www.boost.org/doc/libs/1_65_1/doc/html/stacktrace/configuration_and_build.html) module where it is also very important exactly which library you link against to get the behavior you want.

The current implementation of the Boost Dependency only allows library names as "modules". This was the most reasonable implementation I could come up with. I'm sorry for breaking old builds..

**Comment by nirbheek** [comment]
> The current implementation of the Boost Dependency only allows library names as "modules". This was the most reasonable implementation I could come up with. I'm sorry for breaking old builds..

Makes sense, thanks for doing the research. Could you add a link to this PR to one of the commits (or in the source) once you're done reverting back to the old behaviour? This research should be accessible from the git history, and PRs aren't easy to find from commits especially if we rebase before pushing.

File: mesonbuild/dependencies/misc.py [hunk 16]
```diff
@@ -158,45 +226,86 @@ def detect_version(self):
...
+
+        # globber1 applies to a layout=system installation
+        # globber2 applies to a layout=versioned installation
+        globber1 = 'libboost_*' if self.static else 'boost_*'
+        globber2 = globber1 + '-' + compiler
+        if self.is_multithreading:
+            globber2 = globber2 + '-mt'
+        if self.is_debug:
+            globber2 = globber2 + '-gd'
+        globber2 = globber2 + '-{}'.format(self.version.replace('.','_'))
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[flake8]__
* [E231] missing whitespace after ','

**Comment by NickeZ** [comment]
I'm done, should I squash or do you have any comments?

**Comment by nirbheek** [comment]
LGTM, you can squash the 'noisy' commits, but do keep the descriptive commits that say things like "Improve win support, correctly pick ..."

**Comment by NickeZ** [comment]
Rebased, extended my commit message a bit. I think this is ready to merge!
